<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Weighted least squares (WLS) in scMultiMap — scMultiMap_WLS • scMultiMap</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Weighted least squares (WLS) in scMultiMap — scMultiMap_WLS"><meta name="description" content="This function implements the 2nd step of scMultiMap:
estimate and infer peak-gene association with WLS.
It takes the output from the 1st step, scMultiMap_IRLS as input."><meta property="og:description" content="This function implements the 2nd step of scMultiMap:
estimate and infer peak-gene association with WLS.
It takes the output from the 1st step, scMultiMap_IRLS as input."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMultiMap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/scMultiMap.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"></ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Weighted least squares (WLS) in scMultiMap</h1>

      <div class="d-none name"><code>scMultiMap_WLS.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function implements the 2nd step of scMultiMap:
estimate and infer peak-gene association with WLS.
It takes the output from the 1st step, <em>scMultiMap_IRLS</em> as input.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scMultiMap_WLS</span><span class="op">(</span><span class="va">count_list</span>, <span class="va">seq_depth_list</span>, <span class="va">pairs_df</span>, <span class="va">irls_list</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p>Cell-type-specific mapping of enhancer and target genes from single-cell multimodal data.
Chang Su, Dongsoo Lee, Peng Jin, Jingfei Zhang;
https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1</p>
    </div>
    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-count-list">count_list<a class="anchor" aria-label="anchor" href="#arg-count-list"></a></dt>
<dd><p>A list of two p by n count matrices.
The first element is for gene and the second for peak.</p></dd>


<dt id="arg-seq-depth-list">seq_depth_list<a class="anchor" aria-label="anchor" href="#arg-seq-depth-list"></a></dt>
<dd><p>A list of two length p vectors of sequencing depths.
The first element is for scRNA-seq and the second for scATAC-seq.</p></dd>


<dt id="arg-pairs-df">pairs_df<a class="anchor" aria-label="anchor" href="#arg-pairs-df"></a></dt>
<dd><p>A data frame of candidate peak-gene pairs to evaluate.
The first column is gene name, and the second is peak name.</p></dd>


<dt id="arg-irls-list">irls_list<a class="anchor" aria-label="anchor" href="#arg-irls-list"></a></dt>
<dd><p>A list of two lists generated from the 1st step of scMultiMap,
i.e. <em>scMultiMap_IRLS</em>. The first list contains the results on genes,
the second on peaks.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Whether to print detailed messages. Default to FALSE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.frame with five columns</p><dl><dt>gene</dt>
<dd><p>gene in the peak-gene pair</p></dd>

<dt>peak</dt>
<dd><p>peak in the peak-gene pair</p></dd>

<dt>pval</dt>
<dd><p>p-value</p></dd>

<dt>test_stat</dt>
<dd><p>test statistics</p></dd>

<dt>covar</dt>
<dd><p>estimated covariance</p></dd>


</dl><p>Each row corresponds to one peak-gene pair's results from scMultiMap.
<em>pval</em> and <em>test_stat</em> denote the statistical significance of peak-gene association,
and <em>covar</em> denote the magnitude of association.
The number of rows is equal to the number of pairs in <em>pairs_df</em> present in the Seurat object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">count_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">small_obj</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> peak <span class="op">=</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">small_obj</span><span class="op">[[</span><span class="st">'peak'</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">seq_depth_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">small_obj</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> peak <span class="op">=</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">small_obj</span><span class="op">[[</span><span class="st">'peak'</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">wls_res</span> <span class="op">&lt;-</span> <span class="fu">scMultiMap_WLS</span><span class="op">(</span><span class="va">count_list</span>, <span class="va">seq_depth_list</span>, <span class="va">small_pairs_df</span>, <span class="va">small_irls_list</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "There are 10 unique genes in the peak-gene pairs."</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chang Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

