<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>WLS utility function. — scMultiMap_WLS_by_gene • scMultiMap</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="WLS utility function. — scMultiMap_WLS_by_gene"><meta name="description" content="This function performs WLS for one gene and all peaks in its candidate pairs."><meta property="og:description" content="This function performs WLS for one gene and all peaks in its candidate pairs."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMultiMap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/scMultiMap.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>WLS utility function.</h1>

      <div class="d-none name"><code>scMultiMap_WLS_by_gene.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs WLS for one gene and all peaks in its candidate pairs.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scMultiMap_WLS_by_gene</span><span class="op">(</span></span>
<span>  <span class="va">x_gene</span>,</span>
<span>  <span class="va">X_peak</span>,</span>
<span>  <span class="va">m_gene</span>,</span>
<span>  <span class="va">M_peak</span>,</span>
<span>  <span class="va">w_gene</span>,</span>
<span>  <span class="va">W_peak</span>,</span>
<span>  <span class="va">var_gene</span>,</span>
<span>  <span class="va">Var_peak</span>,</span>
<span>  <span class="va">gene_seq_depth</span>,</span>
<span>  <span class="va">peak_seq_depth</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p>Cell-type-specific mapping of enhancer and target genes from single-cell multimodal data.
Chang Su, Dongsoo Lee, Peng Jin, Jingfei Zhang;</p>
    </div>
    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x-gene">x_gene<a class="anchor" aria-label="anchor" href="#arg-x-gene"></a></dt>
<dd><p>A length n vector of gene counts. n denotes the number of cells.</p></dd>


<dt id="arg-x-peak">X_peak<a class="anchor" aria-label="anchor" href="#arg-x-peak"></a></dt>
<dd><p>A n by p peak count matrix. p denotes the number of peaks.</p></dd>


<dt id="arg-m-gene">m_gene<a class="anchor" aria-label="anchor" href="#arg-m-gene"></a></dt>
<dd><p>A length n vector of gene counts' mean across cells.</p></dd>


<dt id="arg-m-peak">M_peak<a class="anchor" aria-label="anchor" href="#arg-m-peak"></a></dt>
<dd><p>A n by p matrix of peak counts' mean across cells for all peaks.</p></dd>


<dt id="arg-w-gene">w_gene<a class="anchor" aria-label="anchor" href="#arg-w-gene"></a></dt>
<dd><p>A length n vector used to calculate regression weights.</p></dd>


<dt id="arg-w-peak">W_peak<a class="anchor" aria-label="anchor" href="#arg-w-peak"></a></dt>
<dd><p>A n by p matrix used to calculate regression weights.</p></dd>


<dt id="arg-var-gene">var_gene<a class="anchor" aria-label="anchor" href="#arg-var-gene"></a></dt>
<dd><p>A length n vector of gene counts' variance across cells.</p></dd>


<dt id="arg-var-peak">Var_peak<a class="anchor" aria-label="anchor" href="#arg-var-peak"></a></dt>
<dd><p>A n by p matrix of peak counts' variance across cells for all peaks.</p></dd>


<dt id="arg-gene-seq-depth">gene_seq_depth<a class="anchor" aria-label="anchor" href="#arg-gene-seq-depth"></a></dt>
<dd><p>A length n vector of sequencing depths from the RNA modality.</p></dd>


<dt id="arg-peak-seq-depth">peak_seq_depth<a class="anchor" aria-label="anchor" href="#arg-peak-seq-depth"></a></dt>
<dd><p>A length n vector of sequencing depths from the ATAC/peak modality.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.frame with p rows and three columns</p><dl><dt>pval</dt>
<dd><p>p-value</p></dd>

<dt>test_stat</dt>
<dd><p>test statistics</p></dd>

<dt>covar</dt>
<dd><p>estimated covariance</p></dd>


</dl><p>Each row correspond to scMultiMap's WLS result on one gene-peak pair.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chang Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

