<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Iteratively reweighted least squares (IRLS) in scMultiMap — scMultiMap_IRLS • scMultiMap</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Iteratively reweighted least squares (IRLS) in scMultiMap — scMultiMap_IRLS"><meta name="description" content="This function implements the 1st step of scMultiMap:
estimate mean and variance for peaks or genes with IRLS.
The output will be used as input to the 2nd step, scMultiMap_WLS(), to
estimate and infer the association between genes and peaks."><meta property="og:description" content="This function implements the 1st step of scMultiMap:
estimate mean and variance for peaks or genes with IRLS.
The output will be used as input to the 2nd step, scMultiMap_WLS(), to
estimate and infer the association between genes and peaks."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMultiMap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/scMultiMap.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Iteratively reweighted least squares (IRLS) in scMultiMap</h1>

      <div class="d-none name"><code>scMultiMap_IRLS.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function implements the 1st step of scMultiMap:
estimate mean and variance for peaks or genes with IRLS.
The output will be used as input to the 2nd step, scMultiMap_WLS(), to
estimate and infer the association between genes and peaks.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scMultiMap_IRLS</span><span class="op">(</span><span class="va">X</span>, <span class="va">seq_depth</span>, bsample <span class="op">=</span> <span class="cn">NULL</span>, irls <span class="op">=</span> <span class="cn">T</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p>Cell-type-specific mapping of enhancer and target genes from single-cell multimodal data.
Chang Su, Dongsoo Lee, Peng Jin, Jingfei Zhang;</p>
    </div>
    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">X<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A n by p count matrix, where n denotes the number of cells and p denotes the number of genes/peaks.
It can be a length p vector when n=1.</p></dd>


<dt id="arg-seq-depth">seq_depth<a class="anchor" aria-label="anchor" href="#arg-seq-depth"></a></dt>
<dd><p>A length n vector of sequencing depths</p></dd>


<dt id="arg-bsample">bsample<a class="anchor" aria-label="anchor" href="#arg-bsample"></a></dt>
<dd><p>A length n vector of the names of biological sample the cell is from. When all cells are from the same
sample, set to NULL (Default).</p></dd>


<dt id="arg-irls">irls<a class="anchor" aria-label="anchor" href="#arg-irls"></a></dt>
<dd><p>Whether to use IRLS. Default to TRUE. If FALSE, use ordinary least squares.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Whether to print detailed messages. Default to FALSE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of two length p vectors:</p><dl><dt>mu</dt>
<dd><p>estimated mean</p></dd>

<dt>sigma_sq</dt>
<dd><p>estimated variance</p></dd>


</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># The following codes illustrate the use of `scMultiMap_IRLS` with toy datasets</span></span></span>
<span class="r-in"><span><span class="co"># `small_obj` provided in this R package.</span></span></span>
<span class="r-in"><span><span class="va">irls_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">irls_list</span><span class="op">[[</span><span class="st">'gene'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">scMultiMap_IRLS</span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">small_obj</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">small_obj</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">irls_list</span><span class="op">[[</span><span class="st">'peak'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">scMultiMap_IRLS</span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">small_obj</span><span class="op">[[</span><span class="st">'peak'</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">small_obj</span><span class="op">[[</span><span class="st">'peak'</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">irls_list</span><span class="op">[[</span><span class="st">'gene'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="co"># estimated mean and variance for genes</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ mu      : Named num [1:100] 0.1273 0.0341 0.0241 0.0241 0.0216 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "names")= chr [1:100] "MALAT1" "LYZ" "B2M" "ACTB" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ sigma_sq: Named num [1:100] 1.57e-03 2.05e-04 5.67e-05 7.81e-05 2.00e-04 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "names")= chr [1:100] "MALAT1" "LYZ" "B2M" "ACTB" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ theta   : num 5.05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">irls_list</span><span class="op">[[</span><span class="st">'peak'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="co"># estimated mean and variance for peaks</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ mu      : Named num [1:250] 0.00559 0.00359 0.00423 0.00436 0.00282 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "names")= chr [1:250] "chr11-65013281-65014841" "chr11-65026934-65027962" "chr11-65040583-65041780" "chr11-65083457-65084665" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ sigma_sq: Named num [1:250] -5.06e-06 1.65e-06 5.43e-07 3.94e-06 3.56e-06 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "names")= chr [1:250] "chr11-65013281-65014841" "chr11-65026934-65027962" "chr11-65040583-65041780" "chr11-65083457-65084665" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ theta   : num 4.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chang Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

