---
title: "scMultiMap"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

**The vignettes in this package are currently under active development, with more examples coming soon!**

This vignette shows a basic example of applying `scMultiMap` to infer peak-gene associations in cell types with single-cell multimodal data.

# 1. Load packages and data
```{r setup, warning = FALSE, message = FALSE, eval = FALSE}
library(scMultiMap)
library(Signac)
library(Seurat)
```

In this vignette, we use the single cell multiome data (paired scRNA-seq and scATAC-seq) from 10x. The data `processed_seurat_obj_PBMC_10k.rds` can be generated by following these two Signac tutorials: [Joint RNA and ATAC analysis: 10x multiomic](https://stuartlab.org/signac/articles/pbmc_multiomic.html) and [Calling peaks](https://stuartlab.org/signac/articles/peak_calling). The first tutorial is for annotating PBMC cell types, and the second tutorial is for calling peaks by cell type to identify as many cell-type-specific peaks as possible. 

We note that scMultiMap is designed to model **fragment counts**, which are more appropriate for count based modeling and can yield improved performance in downstream analysis compared to read counts. See [Miao and Kim, 2024](https://www.nature.com/articles/s41592-023-02103-7) and [Martens et al., 2024](https://www.nature.com/articles/s41592-023-02112-6) for more discussions. If you plan to use scMultiMap, please make sure the your scATAC-seq data are fragment counts instead of read counts. One recommended way is to use MACS2 to call peaks which generates fragment counts, as shown in [Calling peaks](https://stuartlab.org/signac/articles/peak_calling). If you only have read counts, fragment counts can be estimated following [Martens et al., 2024](https://www.nature.com/articles/s41592-023-02112-6).

Back to this case study. We load the 10x multiome data on PBMC into the R session:
```{r include = FALSE, eval = FALSE}
pbmc <- readRDS('../../../example_data/processed_seurat_obj_PBMC_10k.rds')
```

```{r eval = FALSE, eval = FALSE}
pbmc <- readRDS('processed_seurat_obj_PBMC_10k.rds')
```

# 2. Select the cell types and peak-gene pairs to study

In this example, we focus on CD14 monocytes:

```{r eval = FALSE}
pbmc_cd14 <- subset(pbmc, predicted.id == 'CD14 Mono')
nrow(pbmc_cd14)
```

There are 3,146 cells in this cell type.

We choose to study the top $2{,}000$ highly expressed genes and top $20{,}000$ highly accessible peaks in CD14 Monocytes, and use a cis-region of width 1Mb to define candidate peak-gene pairs.

```{r include=FALSE}
peak_gene_pairs <- read.table('../../../example_data/CD14 Mono_20000_peak_2000_gene_pairs.txt', header=T)
```

```{r eval=FALSE}
peak_gene_pairs <- read.table('CD14 Mono_20000_peak_2000_gene_pairs.txt', header=T)
nrow(peak_gene_pairs)
```

Here, we load 30,635 candidate peak-gene pairs generated based on the criterion above. `peak_gene_pairs` is a dataframe includes 30,635 rows for pairs and two columns for gene and peak names, respectively. Generally, these pairs should be formed and provided by the users.

# 3. Run `scMultiMap` to identify significantly associated peak-gene pairs

```{r eval=FALSE}
start <- Sys.time()
res <- scMultiMap(pbmc_cd14, peak_gene_pairs[,c('gene','peak')], 
                  gene_assay = 'RNA',
                  peak_assay = 'peaks_ct')
end <- Sys.time() 
end - start
```

* Running scMultiMap on 3,146 cells for 30,635 peak-gene pairs will take ~1 minute.
* `res` is a dataframe with association statistics and p-values for each of the 30,635 peak-gene pairs.

**This vignette is currently under active development. More analysis is coming soon!**
