<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>scMultiMap â€¢ scMultiMap</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="scMultiMap">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMultiMap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/scMultiMap.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav"></ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>scMultiMap</h1>
            
      

      <div class="d-none name"><code>scMultiMap.Rmd</code></div>
    </div>

    
    
<p>This vignette shows two examples of applying <code>scMultiMap</code>
to infer peak-gene associations in cell types with single-cell
multimodal data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://changsubiostats.github.io/scMultiMap/">scMultiMap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="toy-example">1. Toy example<a class="anchor" aria-label="anchor" href="#toy-example"></a>
</h2>
<p>To illustrate the basic functionality of scMultiMap, we use a toy
Seurat object <code>small_obj</code> and a data frame of candidate
peak-gene pairs <code>small_pairs_df</code> contained in the R package
to run scMultiMap:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a Seurat object</span></span>
<span><span class="va">small_obj</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 350 features across 100 samples within 2 assays </span></span>
<span><span class="co">#&gt; Active assay: RNA (100 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span>
<span><span class="co">#&gt;  1 other assay present: peak</span></span>
<span></span>
<span><span class="co"># candidate peak-gene pairs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">small_pairs_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;     gene                    peak</span></span>
<span><span class="co">#&gt; 1 MALAT1 chr11-65013281-65014841</span></span>
<span><span class="co">#&gt; 2 MALAT1 chr11-65026934-65027962</span></span>
<span><span class="co">#&gt; 3 MALAT1 chr11-65040583-65041780</span></span>
<span><span class="co">#&gt; 4 MALAT1 chr11-65083457-65084665</span></span>
<span><span class="co">#&gt; 5 MALAT1 chr11-65095679-65096871</span></span>
<span><span class="co">#&gt; 6 MALAT1 chr11-65110172-65110785</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">small_pairs_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 250   2</span></span></code></pre></div>
<p>With these example data, we then run scMultiMap:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMultiMap.html">scMultiMap</a></span><span class="op">(</span><span class="va">small_obj</span>, <span class="va">small_pairs_df</span><span class="op">)</span> <span class="co"># &lt; 0.1 seconds</span></span>
<span><span class="co">#&gt; [1] "Start step 1: IRLS"</span></span>
<span><span class="co">#&gt; [1] "Start IRLS for RNA"</span></span>
<span><span class="co">#&gt; [1] "Start IRLS for peak"</span></span>
<span><span class="co">#&gt; [1] "Start step 2: WLS"</span></span>
<span><span class="co">#&gt; [1] "There are 10 unique genes in the peak-gene pairs."</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt;     gene                    peak       pval  test_stat         covar      padj</span></span>
<span><span class="co">#&gt; 1 MALAT1 chr11-65013281-65014841 0.15215553  1.4319591  3.100282e-05 0.8273097</span></span>
<span><span class="co">#&gt; 2 MALAT1 chr11-65026934-65027962 0.79384145  0.2613256  5.204042e-06 0.9753992</span></span>
<span><span class="co">#&gt; 3 MALAT1 chr11-65040583-65041780 0.68116942  0.4108679  8.623731e-06 0.9472305</span></span>
<span><span class="co">#&gt; 4 MALAT1 chr11-65083457-65084665 0.03236884  2.1398259  4.879823e-05 0.7196890</span></span>
<span><span class="co">#&gt; 5 MALAT1 chr11-65095679-65096871 0.02264609 -2.2793530 -4.308216e-05 0.7196890</span></span>
<span><span class="co">#&gt; 6 MALAT1 chr11-65110172-65110785 0.88810490 -0.1407026 -1.919029e-06 0.9881934</span></span>
<span><span class="co">#&gt;           cor</span></span>
<span><span class="co">#&gt; 1  0.00000000</span></span>
<span><span class="co">#&gt; 2  0.10216095</span></span>
<span><span class="co">#&gt; 3  0.29533510</span></span>
<span><span class="co">#&gt; 4  0.62091902</span></span>
<span><span class="co">#&gt; 5 -0.57664094</span></span>
<span><span class="co">#&gt; 6 -0.08605519</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 250   7</span></span></code></pre></div>
<p>Here, scMultiMap takes a Seurat object with two modalities: RNA and
peak, and tests the associations between 250 peak-gene pairs. We assume
that this Seurat object contains the cells from a specific cell type of
interest, such that the inferred associations are
cell-type-specific.</p>
</div>
<div class="section level2">
<h2 id="pbmc-example-from-10x-multiome-data">2. PBMC example from 10x multiome data<a class="anchor" aria-label="anchor" href="#pbmc-example-from-10x-multiome-data"></a>
</h2>
<p>Next, we use the single cell multiome data (paired scRNA-seq and
scATAC-seq) from 10x. The data
<code>processed_seurat_obj_PBMC_10k.rds</code> can be generated by
following these two Signac tutorials: <a href="https://stuartlab.org/signac/articles/pbmc_multiomic.html" class="external-link">Joint
RNA and ATAC analysis: 10x multiomic</a> and <a href="https://stuartlab.org/signac/articles/peak_calling" class="external-link">Calling
peaks</a>. The first tutorial is for annotating PBMC cell types, and the
second tutorial is for calling peaks by cell type to identify as many
cell-type-specific peaks as possible.</p>
<p><strong>Note:</strong> We note that scMultiMap is designed to model
<strong>fragment counts</strong>, which are more appropriate for count
based modeling and can yield improved performance in downstream analysis
compared to read counts. See <a href="https://www.nature.com/articles/s41592-023-02103-7" class="external-link">Miao and Kim,
2024</a> and <a href="https://www.nature.com/articles/s41592-023-02112-6" class="external-link">Martens et
al., 2024</a> for more discussions. If you plan to use scMultiMap,
please make sure the your scATAC-seq data are fragment counts instead of
read counts. One recommended way is to use MACS2 to call peaks which
generates fragment counts, as shown in <a href="https://stuartlab.org/signac/articles/peak_calling" class="external-link">Calling
peaks</a>. If you only have read counts, fragment counts can be
estimated following <a href="https://www.nature.com/articles/s41592-023-02112-6" class="external-link">Martens et
al., 2024</a>.</p>
<p>Back to this case study. We load the 10x multiome data on PBMC into
the R session:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'processed_seurat_obj_PBMC_10k.rds'</span><span class="op">)</span></span></code></pre></div>
<p><code>pbmc</code> is a Seurat object with a â€˜RNAâ€™ assay for gene
expression data and â€˜peakâ€™ assay for chromatin accessibility data.</p>
<p>In this example, we focus on CD14 monocytes:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_cd14</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">predicted.id</span> <span class="op">==</span> <span class="st">'CD14 Mono'</span><span class="op">)</span></span></code></pre></div>
<p>There are 3,146 cells in this cell type.</p>
<p>We choose to study the top
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>2</mn><mo>,</mo><mn>000</mn></mrow><annotation encoding="application/x-tex">2{,}000</annotation></semantics></math>
highly expressed genes and top
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>20</mn><mo>,</mo><mn>000</mn></mrow><annotation encoding="application/x-tex">20{,}000</annotation></semantics></math>
highly accessible peaks in CD14 Monocytes, and use a cis-region of width
1Mb to define candidate peak-gene pairs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairs_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_top_peak_gene_pairs.html">get_top_peak_gene_pairs</a></span><span class="op">(</span><span class="va">pbmc_cd14</span>, </span>
<span>                                    gene_top<span class="op">=</span><span class="fl">2000</span>, peak_top<span class="op">=</span><span class="fl">20000</span>, </span>
<span>                                    distance <span class="op">=</span> <span class="fl">5e+5</span>,</span>
<span>                                    gene_assay <span class="op">=</span> <span class="st">'RNA'</span>, peak_assay <span class="op">=</span> <span class="st">'peak'</span><span class="op">)</span></span></code></pre></div>
<p>Here, we load 30,635 candidate peak-gene pairs generated based on the
criterion above. <code>peak_gene_pairs</code> is a dataframe includes
30,635 rows for pairs and two columns for gene and peak names,
respectively. Generally, these pairs should be formed and provided by
the users.</p>
<p>Nect, we run <code>scMultiMap</code> to identify significantly
associated peak-gene pairs</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">start</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMultiMap.html">scMultiMap</a></span><span class="op">(</span><span class="va">pbmc_cd14</span>, <span class="va">pairs_df</span>, </span>
<span>                  gene_assay <span class="op">=</span> <span class="st">'RNA'</span>,</span>
<span>                  peak_assay <span class="op">=</span> <span class="st">'peak'</span><span class="op">)</span></span>
<span><span class="va">end</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">end</span> <span class="op">-</span> <span class="va">start</span></span></code></pre></div>
<ul>
<li>Running scMultiMap on 3,146 cells for 30,635 peak-gene pairs will
take ~1 minute.</li>
<li>
<code>res</code> is a dataframe with association statistics and
p-values for each of the 30,635 peak-gene pairs.</li>
</ul>
<p><strong>This vignettes in this package are currently under active
development. More examples are coming soon!</strong></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chang Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
